x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/InternalCategorizationAggregation.java

```diff
@@ -0,0 +1,436 @@
+            BucketKey bucketKey = (BucketKey) o;
+            return Arrays.equals(key, bucketKey.key);
+        }
+
+        @Override
+        public int hashCode() {
+            return Arrays.hashCode(key);
+        }
+
+        public BytesRef[] keyAsTokens() {
+            return key;
+        }
+
+        @Override
+        public int compareTo(BucketKey o) {
+            return Arrays.compare(key, o.key);
+        }
+
+    }
+
+    public static class Bucket extends InternalBucket implements MultiBucketsAggregation.Bucket, Comparable<Bucket> {
+        // Used on the shard level to keep track of sub aggregations
+        long bucketOrd;
+
+        final BucketKey key;
+        final long docCount;
+        InternalAggregations aggregations;
+
+        public Bucket(BucketKey key, long docCount, InternalAggregations aggregations) {
+            this.key = key;
+            this.docCount = docCount;
+            this.aggregations = aggregations;
+        }
+
+        public Bucket(StreamInput in) throws IOException {
+            key = new BucketKey(in);
+            docCount = in.readVLong();
+            aggregations = InternalAggregations.readFrom(in);
+        }
+
+        @Override
+        public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+            builder.startObject();
+            builder.field(CommonFields.DOC_COUNT.getPreferredName(), docCount);
+            builder.field(CommonFields.KEY.getPreferredName());
+            key.toXContent(builder, params);
+            aggregations.toXContentInternal(builder, params);
+            builder.endObject();
+            return builder;
+        }
+
+        BucketKey getRawKey()  {
+            return key;
+        }
+
+        @Override
+        public Object getKey() {
+            return key;
+        }
+
+        @Override
+        public String getKeyAsString() {
+            return key.asString();
+        }
+
+        @Override
+        public long getDocCount() {
+            return docCount;
+        }
+
+        @Override
+        public Aggregations getAggregations() {
+            return aggregations;
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            key.writeTo(out);
+            out.writeVLong(getDocCount());
+            aggregations.writeTo(out);
+        }
+
+        @Override
+        public String toString() {
+            return "Bucket{" + "key=" + getKeyAsString() + ", docCount=" + docCount + ", aggregations=" + aggregations.asMap() + "}\n";
+        }
+
+        @Override
+        public int compareTo(Bucket o) {
+            return key.compareTo(o.key);
+        }
+
+    }
+
+    private final List<Bucket> buckets;
+    protected final int maxUniqueTokens;
+    protected final int similarityThreshold;
+    protected final int maxMatchTokens;
+    protected final int requiredSize;
+    protected final long minDocCount;
```

Comment:
Can these be `private` - every time I see a member variable that isn't private I have to stop reading code and spend time figuring out why. If we can get away with `private`, it'd be nice.

Reply:
The reason is the unmapped deal. Since we inherit from this class, its just way easier to have them protected so that unmapped has them when creating an empty aggregation.

I can add a bunch of getters

Reply:
Nah. I think it's fine. If we have a good reason we can do whatever. Especially as they are `final`.